esp-hal/src/clock/mod.rs

```diff
@@ -659,10 +678,177 @@ impl<'d> RadioClockController<'d> {
         clocks_ll::ble_rtc_clk_init();
     }
 
-    /// Reset the Resolvable Private Address (RPA).
+    /// Reset the Bluetooth Resolvable Private Address (RPA).
     #[instability::unstable]
     #[inline]
     pub fn reset_rpa(&mut self) {
         clocks_ll::reset_rpa();
     }
+    /// Split the radio clock controller into individual modem clock controllers.
+    #[instability::unstable]
+    #[inline]
+    pub fn split(self) -> ModemClockControllers<'d> {
+        ModemClockControllers {
+            #[cfg(wifi)]
+            wifi: WiFiClockController { _phantom: PhantomData },
+            #[cfg(bt)]
+            bt: BtClockController { _phantom: PhantomData },
+            #[cfg(ieee802154)]
+            ieee802154: Ieee802154ClockController { _phantom: PhantomData }
+        }
+    }
+}
+
+#[cfg(any(bt, ieee802154, wifi))]
+#[instability::unstable]
+#[derive(Clone, Copy)]
+#[doc(hidden)]
+/// These are all modems with bitmasks as their representation.
+pub enum Modem {
+    #[cfg(wifi)]
+    WiFi       = 0,
+    #[cfg(bt)]
+    Bt         = 1,
+    #[cfg(ieee802154)]
+    Ieee802154 = 2,
+}
+
+#[cfg(any(bt, ieee802154, wifi))]
+/// Tracks which modems currently request an active PHY clock.
+static PHY_CLOCK_REF_COUNTERS: critical_section::Mutex<[AtomicU8; 3]> = critical_section::Mutex::new([const { AtomicU8::new(0) }; 3]);
```

Comment:
Use `Cell<u8>` please. If you are in a critical section, these don't need to be atomic.